### PR TITLE
Nodes: New features and revisions

### PR DESCRIPTION
**Pull request by sunag**
**Description**

### CacheNode

The new caching system will be the basis for radiance and irradiance flows which should use different caches, it also adds partial support for environment using `Texture2D` in WebGPU but needs `PMREM` to work properly. The cache will also be used in upcoming `SwitchNode` and `LoopNode` features.

API:
```js
const isolatedNode = cache( someNode ); // creates an independent cache
```

### StackNode

API:
```js
const computeShaderNode = new ShaderNode( ( inputs, stack ) => {

	// ...

	const storageElementNode = element( someStorageNode, instanceIndex );

	// StackNode API -> stack.assign( targetNode, sourceNode )

	stack.assign( storageElementNode, valueNode );

} );
```

*This contribution is funded by [Google](https://google.com)*

### CODE REVIEW

File: examples/jsm/nodes/core/NodeBuilder.js [hunk 1]
```diff
@@ -192,6 +210,12 @@ class NodeBuilder {
 
 	}
 
+	getMIPLevelAlgorithmNode( textureNode, levelNode ) {
```
Comment (LeviPesin) [inline examples/jsm/nodes/core/NodeBuilder.js:213]: Why we need this as a method of `NodeBuilder`?

File: examples/jsm/nodes/core/NodeBuilder.js [hunk 2]
```diff
@@ -450,13 +476,11 @@ class NodeBuilder {
 
 	getNodeProperties( node, shaderStage = this.shaderStage ) {
 
-		const nodeData = this.getDataFromNode( this, shaderStage );
-		const constructHash = node.getConstructHash( this );
+		const nodeData = this.getDataFromNode( node, shaderStage );
 
-		nodeData.properties = nodeData.properties || {};
-		nodeData.properties[ constructHash ] = nodeData.properties[ constructHash ] || { outputNode: null };
+		nodeData.properties = nodeData.properties || { outputNode: null };
```
Comment (LeviPesin) [inline examples/jsm/nodes/core/NodeBuilder.js:481]: Can we use `||=` here?

File: examples/jsm/nodes/core/NodeBuilder.js (hunk 1 above)
Comment (sunag) [reply examples/jsm/nodes/core/NodeBuilder.js:213]: `SamplerLevel` can be controlled in [ 0, 1 ] values, for this worked in PBR is necessary an algorithm, I'll remove it from here and move to `EnvironmentNode` as `Context` and use just a linear approach for non-pbr until I finish this PR.

https://github.com/mrdoob/three.js/pull/25041/files#diff-b3dec78c4d4656e9b96d66830127da312a15734c78494a3717e5c3a7d26a5320

**Comment by sunag** [comment]
@mrdoob @LeviPesin `stack.assign( a, b )` instead of `assign( a, b ).build( builder )`, and more stable.

**Pull request merged**